src/linear_algebra/contraction.lean

```diff
@@ -51,4 +64,67 @@ variables {R M N}
   dual_tensor_hom R M N (f ⊗ₜ n) m = (f m) • n :=
 by { dunfold dual_tensor_hom, rw uncurry_apply, refl, }
 
+/-- As a matrix, `dual_tensor_hom` evaluated on a basis element of `M* ⊗ N` is a matrix with a
+single one and zeros elsewhere -/
+theorem to_matrix_dual_tensor_hom
+  {m : Type*} {n : Type*} [fintype m] [fintype n] [decidable_eq m] [decidable_eq n]
+  (bM : basis m R M) (bN : basis n R N) (j : m) (i : n) :
+    to_matrix bM bN (dual_tensor_hom R M N (bM.coord j ⊗ₜ bN i)) = std_basis_matrix i j 1 :=
+begin
+  ext i' j',
+  by_cases hij : (i = i' ∧ j = j');
+  simp [linear_map.to_matrix_apply, finsupp.single_eq_pi_single, hij],
+  rw [and_iff_not_or_not, not_not] at hij, cases hij; simp [hij],
+end
+
+/-- `hom_dual_tensor` is right inverse to `dual_tensor_hom` -/
+@[simp]
+lemma dual_tensor_hom_hom_dual_tensor_apply (f : M →ₗ[R] N):
+  (dual_tensor_hom R M N) ((hom_dual_tensor R M N b) f) = f :=
+begin
+  ext m, nth_rewrite_rhs 0 ←basis.sum_repr b m,
+  simp [hom_dual_tensor],
+end
+
+/-- `hom_dual_tensor` is right inverse to `dual_tensor_hom` -/
+lemma dual_tensor_hom_hom_dual_tensor :
+  (dual_tensor_hom R M N) ∘ₗ (hom_dual_tensor R M N b) = id :=
+by { ext f, simp }
+
+local attribute [ext] tensor_product.ext
+
+lemma hom_dual_tensor_dual_tensor_hom :
+  hom_dual_tensor R M N b ∘ₗ dual_tensor_hom R M N = linear_map.id :=
+begin
+  ext f m,
+  show hom_dual_tensor R M N b (dual_tensor_hom R M N (f ⊗ₜ[R] m)) = f ⊗ₜ[R] m,
+  simp_rw [hom_dual_tensor_apply, dual_tensor_hom_apply f _ m, ←smul_tmul, ←sum_tmul,
+    basis.coe_dual_basis, basis.sum_dual_apply_smul_coord],
+end
+
+/-- If `M` is free, the natural linear map $M^* ⊗ N → Hom(M, N)$ is an equivalence. This function
+provides this equivalence in return for a basis of `M`. -/
+@[simps] noncomputable def dual_tensor_hom_equiv_of_basis :
+  (module.dual R M) ⊗[R] N ≃ₗ[R] M →ₗ[R] N :=
+linear_equiv.of_linear (dual_tensor_hom R M N) (hom_dual_tensor R M N b)
+  (dual_tensor_hom_hom_dual_tensor b) (hom_dual_tensor_dual_tensor_hom b)
+
+@[simp] lemma dual_tensor_hom_equiv_of_basis_to_linear_map :
+  (dual_tensor_hom_equiv_of_basis b : (module.dual R M) ⊗[R] N ≃ₗ[R] M →ₗ[R] N).to_linear_map =
+  dual_tensor_hom R M N :=
+rfl
+
+end comm_ring
+
+section field
```

Comment:
I think that `field` should never be needed to talk about the trace, and that everything can be done for `module.free`.

Reply:
Could you take a look at this suggestion?

Reply:
@antoinelab01 I don't think you ever addressed this comment, so I'm changing back to "awaiting author"

Reply:
Oh no, I just realized that I thought I had responded to the suggestion but my comment was pending all this time. I was wondering why my comments were left unanswered, that explains it... I'm really sorry about that.

Reply:
@riccardobrasca see the above replies. From the docs it looks like we might have had a type class for this (called `finite_free`?) and that maybe it disappeared for some reason. I'm not too familiar with this part of the algebra library.

Reply:
`finite_free` it's just `[module.finite]` and `module.free`. [This](https://leanprover-community.github.io/mathlib_docs/linear_algebra/free_module/finite/basic.html) contains the basic API, for example [module.free.choose_basis_index.fintype](https://leanprover-community.github.io/mathlib_docs/linear_algebra/free_module/finite/basic.html#module.free.choose_basis_index.fintype) says that any basis is finite. We also have results about the rank, [here](https://leanprover-community.github.io/mathlib_docs/linear_algebra/free_module/finite/rank.html).

If we have something for vector space instead of free modules we should really try to generalize the result. I had a planned refactor, but then I started working on other projects...

Reply:
The right thing to do is to delete `finite_dimensional` entirely, and only use `module.free`. So in practice we should move the results of `linear_algebra.finite_dimensional` to `linear_algebra.free_module.something`, but there are lemmas that have to be moved (I don't remember exactly which one) to avoid circular imports.

Reply:
Thanks, I've made the changes